### DISCUSSION TITLE
Lucene geographic queries

### DISCUSSION THREAD

**Discussion by kevinsmgov** [OrchardCMS/OrchardCore#11478] [Q&A]
I have [geo_bounding_box](https://docs.orchardcore.net/en/dev/docs/reference/modules/Spatial/#geo-bounding-box) working but I'm getting null errors trying to use [geo_polygon](https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-geo-polygon-query.html). Has anyone successfully used get_polygon bounded searches?

**Comment by Skrypt** [comment]
Not implemented yet. Feel free to create a Pull Request.

https://github.com/OrchardCMS/OrchardCore/blob/main/src/OrchardCore/OrchardCore.Lucene.Core/QueryProviders/Filters/GeoBoundingBoxFilterProvider.cs

Here is the GeoBoundingBox code. Just need another one for the GeoPolygon.

**Comment by Skrypt** [comment]
ctx.ReadShapeFromWkt() = http://docs.opengeospatial.org/is/18-010r7/18-010r7.html

https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-geo-polygon-query.html

geo_polygon is deprecated in favor of : 

https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-geo-shape-query.html

geo_shape

Here not sure how to create a Polygon with Lucenet.NET other than WKT but here are some Unit Tests examples with WKT.

https://github.com/NightOwl888/Spatial4n/blob/v0.4.1.0/Spatial4n.Tests/shape/NtsGeometryTest.cs

Also, these are geo shapes coordinates. So, we would need a different type of Lucene Indexation for these.

**Reply by kevinsmgov** [reply]
I'm getting the null reference error back from that also. When I pass the query to the query interface of the Orchard Core Admin, I get 

`Invalid query : Object reference not set to an instance of an object.`

This is the query I'm passing
```
{
    "query": {
        "bool": {
            "must": {
                "match_all": {}
            },
            "filter": {
                "geo_shape": {
                    "LocalBusiness.Point": {
                        "shape": {
                            "type": "envelope",
                            "coordinates": [[-118.49466, 34.02285], [-118.48789, 34.01677], [-118.49494, 34.01144], [-118.50198, 34.01746]]
                        },
                        "relation": "within"
                    }
                }
            }
        }
    }
}
```
which appears to match the documentation. I'm trying to find points within the rectangle. (points that are returned with the geo_bounding_box query)

**Reply by kevinsmgov** [reply]
I read your second post but assumed it supplanted the first. I had thought that the documentation that Orchard pointed to was for the lucene.net capabilities, not an service layer on top of Lucene. I'm trying to find the corresponding objects and methods in lucene.net to support an arbitrary polygon (obviously, it is supported as elistic.co has an API for it, but it's not obvious yet).

**Reply by Skrypt** [reply]
What I meant is that geo_polygon ElasticSearch Query is deprecated so we should implement geo_shape instead. That's because we are trying to follow the ElasticSearch implementation.

Of course, nothing is implemented yet. I just looked quickly at some documentation to see how it could be done.
From what I understand WKT could be used to define a country zone because it contains a lot of geo points to form a shape.
Though, for an arbitrary polygon right now I'm not sure how to define it. I will need to read Lucene.NET documentation and figure out how to make one. In the end, that polygon might need to be achieved with a WKT representation.

**Reply by kevinsmgov** [reply]
It looks like the filtering is leveraging the "spatial4n" package (it's not inherent in the lucene.net package) so I'm checking into that.

For context, City of Santa Monica streets, due to our position on the coast, are angled slightly off of 45° from true north. So, to select a portion of the city for filtering (e.g. the "downtown" area) and outline it by the streets we can't use the geo_bounding_box with precision. (looks like your city has a similar issue)

So, if I am able to create a new ILuceneBooleanFilterProvider that surfaces the geo_shape filter, can I declare it in my project to extend the search capabilities or do I have to wait until it's included in Orchard Core in order to put it into production?

**Reply by Skrypt** [reply]
I think you could extend OrchardCore.Lucene.Core if you want to have this working with a Nuget packages solution. But, knowing that this could be something that other people might need it would be worth creating a proper pull request on Orchard Core to add the filter so that everyone can use it.

Though, depending on your current website solution it would be safer for you to create an extension of OrchardCore.Lucene.Core so that you don't wait on the pull request to be merged. Also, if we eventually merge that pull request then you need to start using the latest "dev" Nuget packages on your solution or wait for a proper next release if you feel concerned about using "in dev" bits.

**Reply by Skrypt** [reply]
And yes for spatial4n it is used for every Spatial query in Lucene.NET so far from what I've seen.

**Reply by kevinsmgov** [reply]
I'll need to actually create the working filter filter first before I worry about packages, etc.

Can you point me to an example of how to extend an OC nuget package (I'm assuming it's not just as simple as declaring
`services.AddSingleton<ILuceneBooleanFilterProvider, GeoShapeFilterProvider>();` in my startup file)

**Reply by Skrypt** [reply]
Normally you should extend the AddLuceneQueries() method but I just found out that this method is "public static". And yes what you are saying might just work. It will be enabled from your custom module though, not from the Lucene module.

**Reply by kevinsmgov** [reply]
Excellent, I'll give it a shot.

**Reply by kevinsmgov** [reply]
The good news is that it was easy to extend the Lucene nuget by just adding that line to my module startup. The bad news is that the default spatial4n library does not support the POLYGON WKT type (their example docs use "ENVELOPE" which is just another type of rectangle shape definition). There is an NtsWktShapeParser in the spatial4n repository that adds POLYGON parsing support but it is not showing up in the nuget package. Still looking for an alternative.

**Reply by Skrypt** [reply]
https://lucenenet.apache.org/docs/4.8.0-beta00016/api/spatial/Lucene.Net.Spatial.html

SpatialContext.Geo.WktShapeParser

When I'm looking at the documentation they say it supports NTS.

**Reply by Skrypt** [reply]
Ok, I see what you mean now `Lucene.Net.Contrib.Spatial.NTS` is not available in .NET Core but in the documentation they say : 

> The spatial module is new to Lucene.NET 4, replacing the old "Lucene.Net.Contrib" module that came before it.

**Reply by kevinsmgov** [reply]
I'm comparing https://github.com/Spatial4n/Spatial4n/blob/release/v0.4.1.1/src/Spatial4n/IO/WktShapeParser.cs (which is what I believe Lucene.Net.Spatial is using) versus https://github.com/Spatial4n/Spatial4n/blob/release/v0.4.1.1/src/Spatial4n/IO/Nts/NtsWktShapeParser.cs which is where POLYGON parsing was added. I don't see the second namespace visible in the nuget package that Orchard imports.

**Reply by Skrypt** [reply]
Ok so now this is something that needs to be implemented in Lucene.NET.

**Reply by Skrypt** [reply]
I'm surprised it is not implemented as the Spatial4n repository is managed by the same guy than Lucene.NET. The idea is that Lucene.Net.Contrib should now be mostly part of Lucene.NET 4.x

**Reply by Skrypt** [reply]
I think I'm confused because it seems to work.

```C#
using System;
using System.Linq;
using System.Text.RegularExpressions;
using Lucene.Net.Queries.Function;
using Lucene.Net.Search;
using Lucene.Net.Spatial.Prefix;
using Lucene.Net.Spatial.Prefix.Tree;
using Lucene.Net.Spatial.Queries;
using Lucene.Net.Spatial.Util;
using Newtonsoft.Json.Linq;
using Spatial4n.Context;
using Spatial4n.Context.Nts; // Here I added this using
using Spatial4n.Distance;
using Spatial4n.Shapes;
using Spatial4n.Util;

namespace OrchardCore.Lucene.QueryProviders.Filters
{
    public class GeoDistanceFilterProvider : ILuceneBooleanFilterProvider
    {
        public FilteredQuery CreateFilteredQuery(ILuceneQueryService builder, LuceneQueryContext context, string type, JToken filter, Query toFilter)
        {
            if (type != "geo_distance")
            {
                return null;
            }

            if (!(toFilter is BooleanQuery booleanQuery))
            {
                return null;
            }

            var queryObj = filter as JObject;

            if (queryObj.Properties().Count() != 2)
            {
                return null;
            }

            var ctx = SpatialContext.Geo;
            var ctxNTS = NtsSpatialContext.Geo; // Here I'm using a NtsSpatialContext instead of SpatialContext
            var parser = ctxNTS.WktShapeParser;

            var maxLevels = 11; //results in sub-meter precision for geohash

            //  This can also be constructed from SpatialPrefixTreeFactory
            SpatialPrefixTree grid = new GeohashPrefixTree(ctx, maxLevels);

            JProperty distanceProperty = null;
            JProperty geoProperty = null;

            foreach (var jProperty in queryObj.Properties())
            {
                if (jProperty.Name.Equals("distance", StringComparison.Ordinal))
                {
                    distanceProperty = jProperty;
                }
                else
                {
                    geoProperty = jProperty;
                }
            }

            if (distanceProperty == null || geoProperty == null)
            {
                return null;
            }

            var strategy = new RecursivePrefixTreeStrategy(grid, geoProperty.Name);

            if (!TryParseDistance((string)distanceProperty.Value, out var distanceDegrees))
            {
                return null;
            }

            if (!TryGetPointFromJToken(geoProperty.Value, out var point))
            {
                return null;
            }

            var circle = ctx.MakeCircle(point.X, point.Y, distanceDegrees);

            var args = new SpatialArgs(SpatialOperation.Intersects, circle);

            var spatialQuery = strategy.MakeQuery(args);
            var valueSource = strategy.MakeRecipDistanceValueSource(circle);
            var valueSourceFilter = new ValueSourceFilter(new QueryWrapperFilter(spatialQuery), valueSource, 0, 1);

            booleanQuery.Add(new FunctionQuery(valueSource), Occur.MUST);

            return new FilteredQuery(booleanQuery, valueSourceFilter);
        }

        private static bool TryParseDistance(string distanceValue, out double distanceDegrees)
        {
            distanceDegrees = -1;

            var distanceString = Regex.Match(distanceValue, @"^((\d+(\.\d*)?)|(\.\d+))").Value;

            if (String.IsNullOrEmpty(distanceString))
            {
                return false;
            }

            var distanceUnits = distanceValue.Substring(distanceString.Length, distanceValue.Length - distanceString.Length).ToLower();

            if (!Double.TryParse(distanceString, out var distance))
            {
                return false;
            }

            switch (distanceUnits)
            {
                case "mi":
                case "miles":
                    distanceDegrees = DistanceUtils.Dist2Degrees(distance, DistanceUtils.EarthMeanRadiusMiles);
                    return true;
                case "km":
                case "kilometers":
                    distanceDegrees = DistanceUtils.Dist2Degrees(distance, DistanceUtils.EarthMeanRadiusKilometers);
                    return true;
                case "ft":
                case "feet":
                    distanceDegrees = DistanceUtils.Dist2Degrees(distance / 5280, DistanceUtils.EarthMeanRadiusMiles);
                    return true;
                case "yd":
                case "yards":
                    distanceDegrees = DistanceUtils.Dist2Degrees(distance / 1760, DistanceUtils.EarthMeanRadiusMiles);
                    return true;
                case "in":
                case "inch":
                    distanceDegrees = DistanceUtils.Dist2Degrees(distance / 63360, DistanceUtils.EarthMeanRadiusMiles);
                    return true;
                case "m":
                case "meters":
                    distanceDegrees = DistanceUtils.Dist2Degrees(distance / 1000, DistanceUtils.EarthMeanRadiusKilometers);
                    return true;
                case "cm":
                case "centimeters":
                    distanceDegrees = DistanceUtils.Dist2Degrees(distance / 100000, DistanceUtils.EarthMeanRadiusKilometers);
                    return true;
                case "mm":
                case "millimeters":
                    distanceDegrees = DistanceUtils.Dist2Degrees(distance / 1000000, DistanceUtils.EarthMeanRadiusKilometers);
                    return true;
                case "nm":
                case "nmi":
                case "nauticalmiles":
                    distanceDegrees = DistanceUtils.Dist2Degrees(distance * 1.852, DistanceUtils.EarthMeanRadiusKilometers);
                    return true;
            }

            return false;
        }

        private static bool TryGetPointFromJToken(JToken geoToken, out IPoint point)
        {
            point = null;

            var ctx = SpatialContext.Geo;

            switch (geoToken.Type)
            {
                case JTokenType.String:
                    var geoStringValue = geoToken.ToString();

                    var geoStringSplit = geoStringValue.Split(',');

                    if (geoStringSplit.Length == 2)
                    {
                        if (!Double.TryParse(geoStringSplit[0], out var lat))
                        {
                            return false;
                        }

                        if (!Double.TryParse(geoStringSplit[1], out var lon))
                        {
                            return false;
                        }

                        point = new Point(lon, lat, ctx);
                        return true;
                    }

                    point = GeohashUtils.Decode(geoStringValue, ctx);
                    return true;
                case JTokenType.Object:
                    var geoPointValue = (JObject)geoToken;

                    if (!geoPointValue.ContainsKey("lon") || !geoPointValue.ContainsKey("lat"))
                    {
                        return false;
                    }

                    point = new Point(geoPointValue["lon"].Value<double>(), geoPointValue["lat"].Value<double>(), ctx);
                    return true;
                case JTokenType.Array:
                    var geoArrayValue = (JArray)geoToken;

                    if (geoArrayValue.Count != 2)
                    {
                        return false;
                    }

                    point = new Point(geoArrayValue[0].Value<double>(), geoArrayValue[1].Value<double>(), ctx);

                    return true;
                default: throw new ArgumentException("Invalid geo point representation");
            }
        }
    }
}
```

**Reply by kevinsmgov** [reply]
What does the package import in your CSPROJ file look like for spatial4n?

**Reply by Skrypt** [reply]
It's the same Lucene.Net.Spatial 4.8.0-beta00016 package. It imports the proper Spatial4n package.

**Reply by Skrypt** [reply]
![image](https://user-images.githubusercontent.com/3228637/161620316-b2f470fc-0b46-4ed1-a392-4c98188dbe4f.png)

**Reply by kevinsmgov** [reply]
I'm close but somewhere I picked up a reference to
`Spatial4n.Core, Version=0.4.1.0` (which doesn't exist)
and the site stopped running. I'll let you know when I get that figured out.

**Reply by kevinsmgov** [reply]
What version of Orchard Core are you showing me source from? When I look at the filters, their namespace references are to Spatial4n.Core.Context, not Spatial4n.Context

(e.g. GeoDistanceFilterProvider)
```
using Spatial4n.Core.Context;
using Spatial4n.Core.Distance;
using Spatial4n.Core.Shapes;
using Spatial4n.Core.Shapes.Impl;
using Spatial4n.Core.Util;
```

**Reply by Skrypt** [reply]
The latest main branch of the source code. We may have upgraded to 0.4.1.1 but I doubt it would make a difference else we are just not lucky.

**Reply by Skrypt** [reply]
```C#
using Spatial4n.Context;
using Spatial4n.Context.Nts;
using Spatial4n.Distance;
using Spatial4n.Shapes;
using Spatial4n.Util;
```

**Reply by kevinsmgov** [reply]
I'm running Orchard 1.3.0 and it's using spatial4n 0.4.1.1 and the source from Orchard Core does not use those namespaces

**Reply by kevinsmgov** [reply]
Ahh, I see that main branch has the different namespaces.

**Reply by kevinsmgov** [reply]
It looks like this conflict between 0.4.1.0 and 0.4.1.1 in the spatial4n library is from Orchard Core 1.3.0 (what I'm developing against to try and get a production solution) using Lucene.net 4.8.0-beta00015 and the Orchard Core main branch (where you have access to NTS) using 4.8.0-beta00016

**Reply by kevinsmgov** [reply]
> It looks like this conflict between 0.4.1.0 and 0.4.1.1 in the spatial4n library is from Orchard Core 1.3.0 (what I'm developing against to try and get a production solution) using Lucene.net 4.8.0-beta00015 and the Orchard Core main branch (where you have access to NTS) using 4.8.0-beta00016. 

If I try including Lucene.net 4.8.0-beta00016 in my module to get NTS, the site stops working due to the version conflict.

**Comment by Skrypt** [comment]
I think you will need to use the development NuGet feed and use the latest bits from the main branch.

[![Cloudsmith](https://api-prd.cloudsmith.io/badges/version/orchardcore/preview/nuget/OrchardCore.Application.Cms.Targets/latest/x/?render=true&badge_token=gAAAAABey9hKFD_C-ZIpLvayS3HDsIjIorQluDs53KjIdlxoDz6Ntt1TzvMNJp7a_UWvQbsfN5nS7_0IbxCyqHZsjhmZP6cBkKforo-NqwrH5-E6QCrJ3D8%3D)](https://cloudsmith.io/~orchardcore/repos/preview/packages/detail/nuget/OrchardCore.Application.Cms.Targets/latest/)

**Reply by kevinsmgov** [reply]
Looks like I'll need to hold for now. I'll loop back on this when OC 1.4.x is released and add the filter at that time.

The filter I created is specific to my search so it would not have been a generic WKT enhancement to the JSON structure. However, we have actually been looking at leveraging elastic.co's service and, from what I can see, their geographic search has the same limitation (i.e. no POLYGON boundary search) so I'll see what their plans are for enhancing their service (that way we can stay in sync with their developing documentation).

In the mean time, I might be able to get some intermediate results with a buffered linestring (i.e. running lines down the centerlines of streets with the buffer large enough to encompass business on either side).

**Reply by Skrypt** [reply]
ElasticSearch OC module is currently a work in progress. We have a proof of concept pull request only. It will eventually work but for now, Lucene is what we support only. I'm supposed to work on the ElasticSearch module as soon as I can but of course, you could leverage your own custom module for your specific needs too that uses ElasticSearch in the meantime.

**Reply by kevinsmgov** [reply]
I looked at Lucene.net code on the geo searches and it looks like, whatever the shape that is passed to it, gets surrounded but a rectangular bounding box. I'm a little concerned about the ability to actually do a true polygon contained search with Lucene.

**Reply by Skrypt** [reply]
Normally, the rectangular bounding box is there to limit the number of locations it needs to make calculations from. For example, on a distance search, you want to limit the number of locations where it should calculate the distance from a center point. Else, if you are not limiting the number of locations it will take every location in your index which might take way more time to calculate and return results from.

Here the "Bounding Box" on a radius distance search for example; surrounds the inner circle on which it needs to evaluate a distance based on an Haversine formula. This Harversine formula is then only applied on every locations situated inside that "Bounding Box". So, the "Bounding Box" is used for performance reasons.

It is explained here : 

https://docs.orchardcore.net/en/dev/docs/reference/modules/Spatial/

**Reply by kevinsmgov** [reply]
OK, the BufferedLineString should be effective for the near term. I have a GeoShapeFilterProvider that is scaffolded to make the query work (see below). I'll need to do more work on its parser to manage all of the supported GeoShape geometries before I can submit it to Orchard Core.

![image](https://user-images.githubusercontent.com/5798763/162014785-706631cb-7c5d-4c5c-ac6a-bca8e5596ab0.png)

**Reply by Skrypt** [reply]
Nice! 😄 

Here the BufferedLine should not require any Bounding Box since it already has a "buffer" that should be adjustable. But that works only for a straight line (road). I'm guessing that now you want a Polygon because a road could be also a pronounced curved line.

**Reply by kevinsmgov** [reply]
Actually, the polygon would be for our downtown area (or others) where a series of blocks are outlined. My approach with the LINESTRING is to serpentine through the downtown area. It works but it is a kluge.

So, if I want to make my geo_shape driver generic enough to submit to Orchard Core with a PR, I'll need to figure out how the query DSL would support me passing in my LINESTRING and also the BUFFER value. 

Spatial4n's WktShapeParser (in the version that OC 1.3.0 supports) will process the following WKT geometries (I copied this from their source code documentation):

 - POINT
 - MULTIPOINT
 - ENVELOPE (strictly isn't WKT but is defined by OGC's [http://docs.geoserver.org/stable/en/user/tutorials/cql/cql_tutorial.html](http://docs.geoserver.org/stable/en/user/tutorials/cql/cql_tutorial.html) Common Query Language(CQL)</a>)
 - LINESTRING
 - MULTILINESTRING
 - GEOMETRYCOLLECTION
 - BUFFER (non-standard Spatial4n operation)

So, I would need to traverse the incoming JObject structure and refactor the passed geometries into WKT. However, to add the BUFFER value in, I would be creating a custom variation of the query DSL to support it (and, at that point, it doesn't even need to be parsed as the buffer value would already be a numeric JToken). Any suggestions on altering the DSL?

**Reply by kevinsmgov** [reply]
(or, I can leave my task specific version of the filter in place and loop back to this when our project updates to OC 1.4.0 and implement the POLYGON shape in a filter driver intended for an OC PR and not bother with buffered shapes - assuming that elasticsearch's query DSL never standardizes on buffered shapes)

**Reply by Skrypt** [reply]
If we are altering the DSL then you can add options to it as long as it covers what it has. Also, we need to document these variations.

**Reply by kevinsmgov** [reply]
Looking at Elastic's Lucene implementation, they do have a lot more features enabled. Is the ElasticSearch interface going to be like Orchard's Lucene integration (when an item is published with fields marked for indexing, make an API call to ElasticSearch to update their index)? Or, will it just be an interface to their service for the site search feature (assuming you have configured their service to crawl your site)? (also, is there another place to discuss ElasticSearch integration?)

**Reply by Skrypt** [reply]
This is the Pull Request for ElasticSearch.

https://github.com/OrchardCMS/OrchardCore/pull/11052

It will work just like the Lucene module works except that it will use ElasticSearch / OpenSearch distributed services. Of course, ElasticSearch uses the Java version of Lucene so it is more mature than Lucene.NET as of now. That's the main reason why there are fewer features.

You will be able to use Lucene and ElasticSearch / OpenSearch at the same time. Your Queries will target either Lucene or ElasticSearch in that case. That's a good way to be able to have both running while migrating progressively to one or the other.
